candle-core/src/metal_backend.rs

```diff
@@ -678,33 +680,33 @@ impl BackendStorage for MetalStorage {
+                ("uceil", DType::F32) => strided::ceil::FLOAT,
+                ("ucos", DType::F16) => strided::cos::HALF,
                 ("ucos", DType::F32) => strided::cos::FLOAT,
-                ("usin", DType::F32) => strided::sin::FLOAT,
-                ("usqr", DType::F32) => strided::sqr::FLOAT,
-                ("usqrt", DType::F32) => strided::sqrt::FLOAT,
-                ("uneg", DType::F32) => strided::neg::FLOAT,
-                ("uexp", DType::F32) => strided::exp::FLOAT,
-                ("ulog", DType::F32) => strided::log::FLOAT,
-                ("ugelu", DType::F32) => strided::gelu::FLOAT,
-                ("ugelu_erf", DType::F32) => strided::gelu_erf::FLOAT,
+                ("uerf", DType::F16) => strided::erf::HALF,
                 ("uerf", DType::F32) => strided::erf::FLOAT,
-                ("uceil", DType::F32) => strided::ceil::FLOAT,
+                ("uexp", DType::F16) => strided::exp::HALF,
+                ("uexp", DType::F32) => strided::exp::FLOAT,
+                ("ufloor", DType::F16) => strided::floor::HALF,
                 ("ufloor", DType::F32) => strided::floor::FLOAT,
+                ("ugelu_erf", DType::F16) => strided::gelu_erf::HALF,
+                ("ugelu_erf", DType::F32) => strided::gelu_erf::FLOAT,
+                ("ugelu", DType::F16) => strided::gelu::HALF,
+                ("ugelu", DType::F32) => strided::gelu::FLOAT,
+                ("ulog", DType::F16) => strided::log::HALF,
+                ("ulog", DType::F32) => strided::log::FLOAT,
+                ("uneg", DType::F16) => strided::neg::HALF,
+                ("uneg", DType::F32) => strided::neg::FLOAT,
+                ("uround", DType::F16) => strided::round::HALF,
                 ("uround", DType::F32) => strided::round::FLOAT,
-                ("ucos", DType::F16) => strided::cos::HALF,
                 ("usin", DType::F16) => strided::sin::HALF,
+                ("usin", DType::F32) => strided::sin::FLOAT,
                 ("usqr", DType::F16) => strided::sqr::HALF,
+                ("usqr", DType::F32) => strided::sqr::FLOAT,
                 ("usqrt", DType::F16) => strided::sqrt::HALF,
-                ("uneg", DType::F16) => strided::neg::HALF,
-                ("uexp", DType::F16) => strided::exp::HALF,
-                ("ulog", DType::F16) => strided::log::HALF,
-                ("ugelu", DType::F16) => strided::gelu::HALF,
-                ("ugelu_erf", DType::F16) => strided::gelu_erf::HALF,
-                ("uerf", DType::F16) => strided::erf::HALF,
-                ("uceil", DType::F16) => strided::ceil::HALF,
-                ("ufloor", DType::F16) => strided::floor::HALF,
-                ("uround", DType::F16) => strided::round::HALF,
-                (name, dtype) => crate::bail!("Match {name} - {dtype:?}"),
+                ("usqrt", DType::F32) => strided::sqrt::FLOAT,
+                (name, dtype) => crate::bail!("No binary kernel in metal for {name} - {dtype:?}"),
```

Comment:
Two cents here, this is "strided unary", not "binary" :-)

Reply:
Tried to improve more generally all these in https://github.com/huggingface/candle/pull/1491.

Reply:
Ooops. Thanks! I copied the messages you proposed in your PR to avoid merge conflicts.